File: cpp/swapnodesinpairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (!head or !head->next) return head;
        auto v1 = head;
        auto v2 = head->next;
        auto root = v2;
        auto pre = new ListNode(0);
        while (v2) {
            auto tmp = v2->next;
            v2->next = v1;
            v1->next = tmp;
            pre->next = v2;
            pre = v1;
            v1 = tmp;
            v2 = tmp? tmp->next:NULL;
        }
        return root;
    }
};

File: python/swapnodesinpairs.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head
        cur = head
        root = head.next
        pre = self
        while cur and cur.next:
            tmp = cur.next.next
            cur.next.next = cur
            pre.next = cur.next
            cur.next = tmp
            pre = cur
            cur = tmp
        
        return root
            
